source/agora/test/Transaction.d

```diff
@@ -0,0 +1,264 @@
+{
+    /// Store of previously processed transactions
+    private Transaction[Hash] txs;
+
+    /*******************************************************************************
+
+        Save a `Transaction`
+
+        Params:
+            hash = `Hash` of `Transaction`
+            tx = Type of `Transaction`
+
+    *******************************************************************************/
+
+    public void saveTransaction (const Hash hash, Transaction tx)
+    {
+        this.txs[hash] = tx;
+    }
+
+    /*******************************************************************************
+
+        Find a output of transaction
+
+        Params:
+            hash = `Hash` of `Transaction`
+            index = Index of the `Output` in the `Transaction`
+
+        Return:
+            `Output` in the `Transaction`
+
+    *******************************************************************************/
+
+    public Output findTxOutput(const Hash hash, const uint index)
+    {
+        if (hash in this.txs)
+        {
+            auto tx = this.txs[hash];
+            if (index < tx.outputs.length)
+            {
+                return tx.outputs[index];
+            }
+            else
+            {
+                return Output();
+            }
+        }
+        else
+        {
+            return Output();
+        }
+    }
+
+    /*******************************************************************************
+
+        Get sum of `Input`
+
+        Params:
+            tx = `Transaction`
+
+        Return:
+            Sum of `Input` in the `Transaction`
+
+    *******************************************************************************/
+
+    public long getInputSum (const Transaction tx)
+    {
+        long sum = 0;
+        Output txOut;
+
+        foreach (idx, ref txIn; tx.inputs)
+        {
+            txOut = findTxOutput(txIn.previous, txIn.index);
+            sum += txOut.value;
+        }
+
+        return sum;
+    }
+
+    /*******************************************************************************
+
+        Get sum of `Output`
+
+        Params:
+            tx = `Transaction`
+
+        Return:
+            Sum of `Output` in the `Transaction`
+
+    *******************************************************************************/
+
+    public long getOutputSum (const Transaction tx)
+    {
+        long sum = 0;
+
+        foreach (idx, ref txOut; tx.outputs)
+        {
+            sum += txOut.value;
+        }
+
+        return sum;
+    }
+
+    /*******************************************************************************
+
+        Get result of transaction verification
+
+        Params:
+            tx = `Transaction`
+
+        Return:
+            Return true if this transaction is verified.
+
+    *******************************************************************************/
+
+    public bool verification (const Transaction tx)
+    {
+        if (tx.inputs.length <= 0)
```

Comment:
The input and outputs also need to be validated so they're not negative.

Reply:
Yes, it is. I made a mistake.